Counterexamples/MapFloor.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+
+! This file was ported from Lean 3 source module map_floor
+! leanprover-community/mathlib commit 328375597f2c0dd00522d9c2e5a33b6a6128feeb
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.Order.Hom.Ring
+import Mathlib.Data.Polynomial.Reverse
+
+/-!
+# Floors and ceils aren't preserved under ordered ring homomorphisms
+
+Intuitively, if `f : α → β` is an ordered ring homomorphism, then floors and ceils should be
+preserved by `f` because:
+* `f` preserves the naturals/integers in `α` and `β` because it's a ring hom.
+* `f` preserves what's between `n` and `n + 1` because it's monotone.
+
+However, there is a catch. Potentially something whose floor was `n` could
+get mapped to `n + 1`, and this has floor `n + 1`, not `n`. Note that this is at most an off by one
+error.
+
+This pathology disappears if you require `f` to be strictly monotone or `α` to be archimedean.
+
+## The counterexample
+
+Consider `ℤ[ε]` (`IntWithEpsilon`), the integers with infinitesimals adjoined. This is a linearly
+ordered commutative floor ring (`IntWithEpsilon.linearOrderedCommRing`,
+`IntWithEpsilon.floorRing`).
+
+The map `f : ℤ[ε] → ℤ` that forgets about the epsilons (`IntWithEpsilon.forgetEpsilons`) is an
+ordered ring homomorphism.
+
+But it does not preserve floors (nor ceils) as `⌊-ε⌋ = -1` while `⌊f (-ε)⌋ = ⌊0⌋ = 0`
+(`IntWithEpsilon.forgetEpsilons_floor_lt`, `IntWithEpsilon.lt_forgetEpsilons_ceil`).
+-/
+
+
+namespace Counterexample
+
+noncomputable section
+
+open Function Int Polynomial
+
+open scoped Polynomial
+
+/-- The integers with infinitesimals adjoined. -/
+def IntWithEpsilon :=
+  ℤ[X] deriving Nontrivial
+#align counterexample.int_with_epsilon Counterexample.IntWithEpsilon
+
+local notation "ℤ[ε]" => IntWithEpsilon
+
+local notation "ε" => (X : ℤ[ε])
+
+namespace IntWithEpsilon
+
+instance nontrivial : Nontrivial IntWithEpsilon := inferInstance
+
+-- Porting note: `inhabited` and `commRing` were `deriving` instances in mathlib3
+instance commRing : CommRing IntWithEpsilon := Polynomial.commRing
```

Comment:
I think a cleaner way to write these is
```
instance commRing : CommRing IntWithEpsilon := inferInstanceAs <| CommRing ℤ[X]
```
rather than using the explicit name of the instance.

It's not worth changing it (and it will become unnecessary once there is a deriving handler), but I thought I'd point this out for the future.